cdist/core/cdist_type.py

```diff
@@ -204,11 +204,13 @@ def parameter_defaults(self):
             try:
                 defaults_dir = os.path.join(self.absolute_path, "parameter", "default")
                 for name in os.listdir(defaults_dir):
-                    with open(os.path.join(defaults_dir, name)) as fd:
-                        defaults[name] = fd.read().strip()
+                    try:
+                        with open(os.path.join(defaults_dir, name)) as fd:
+                            defaults[name] = fd.read().strip()
+                    except EnvironmentError:
+                        pass  # Swallow errors raised by open() or read()
             except EnvironmentError:
```

Comment:
Hello @telmich,

Sure, I do understand it could be confusing at first. I hope this will clarify it (let me know if it's still blurry :P ):

Originally the `try ... except EnvironmentError` catches the following errors:
 * `os.listdir(defaults_dir)` where `defaults_dir` doesn't exist
 * `open(file)` where `file` is actually a directory
 * `fd.read()` where `fd` is not readable by the user
 * (other errors that I'm not aware of on `listdir()`, `open()` and `read()`)

The bug I encountered was when there's a directory will be given to `open()`, because the original `try ... except EnvironmentError` is outside the loop, it stops iterating, and keep going.
I added an inner `try ... except` block so that if an `open()` or a `read()` fails for any reason, the loop would keep going. (I added tests for this) And I kept the outer `try ... except` for `os.listdir()`.

Does it make sense?

For the `EnvironmentError`, I didn't notice that, since [`EnvironmentError` was kept for backward compatibility](https://docs.python.org/3/library/exceptions.html#EnvironmentError). If you want to be compatible python 3.2 we should keep `EnvironmentError`, otherwise I think we should rename it to `OSError`.

Let me know if you want me to modify/fix my pull request in any way.

Thank you.